### PR TITLE
【Hackathon 4th No.102】给AutoConverter增加新的模型组网的支持 AlbertModel

### PR DESCRIPTION
**Pull request by megemini**
### PR types

New features

### PR changes

APIs

### Description

【Hackathon 4th No.102】给AutoConverter增加新的模型组网的支持 AlbertModel

Hackathon 4th No.102 这个任务里面有5个模型，我计划每个模型单独提PR，这个PR是处理 albert 模型。

- 使用的测试模型 ``hf-internal-testing/tiny-random-AlbertModel``

反馈几个问题：

- ``AlbertForQuestionAnswering`` 模型无法测试，``transformers`` 的这个模型与 ``paddlenlp`` 有点区别：
  ``` python
  class AlbertForQuestionAnswering(AlbertPreTrainedModel):
      _keys_to_ignore_on_load_unexpected = [r"pooler"]
  
      def __init__(self, config: AlbertConfig):
          super().__init__(config)
          self.num_labels = config.num_labels
  
          self.albert = AlbertModel(config, add_pooling_layer=False) # 注意这里
          self.qa_outputs = nn.Linear(config.hidden_size, config.num_labels)
  
          # Initialize weights and apply final processing
          self.post_init()  
  ```
  ``transformers`` 的 ``albert`` 没有添加 ``pooling`` 层，而 ``paddle`` 是有的，导致无法做name mapping。
  请问这个要怎么处理？

- ``AlbertForMaskedLM`` 与 ``AlbertForPretraining`` 参考 ``BertModel`` 同样无法测试。

@wj-Mcat 请评审，谢谢！：）

### CODE REVIEW

**Comment by megemini** [comment]
刚看了一下 PaddleNLP-CI 失败的原因，好象是别的测试用例出错了？

**Comment by wj-Mcat** [comment]
> 刚看了一下 PaddleNLP-CI 失败的原因，好象是别的测试用例出错了？

是得，这个你暂时不用管。

**Comment by wj-Mcat** [comment]
> transformers 的 albert 没有添加 pooling 层，而 paddle 是有的，导致无法做name mapping。请问这个要怎么处理？

transformers 源码中也是通过`add_pooling_layer`来控制是否添加pooler，所以这里**有可能有**pooler 层，这里你暂时加上pooler 的映射，只不过会抛 warning 。

AlbertForMaskedLM 与 AlbertForPretraining 也是同样来处理。

File: paddlenlp/transformers/albert/modeling.py [hunk 1]
```diff
@@ -357,6 +358,118 @@ class AlbertPretrainedModel(PretrainedModel):
     pretrained_init_configuration = ALBERT_PRETRAINED_INIT_CONFIGURATION
     pretrained_resource_files_map = ALBERT_PRETRAINED_RESOURCE_FILES_MAP
 
+    @classmethod
+    def _get_name_mappings(cls, config: AlbertConfig) -> List[StateDictNameMapping]:
+        mappings: list[StateDictNameMapping] = []
```
Comment (wj-Mcat) [inline paddlenlp/transformers/albert/modeling.py]: 这个变量定义在这里没有用到，故可以删除。

File: tests/transformers/albert/test_modeling.py [hunk 2]
```diff
@@ -337,6 +340,148 @@ def test_model_from_pretrained(self):
...
+                )
+            )
+
+    @parameterized.expand(
+        [
+            ("AlbertModel",),
+            # ("AlbertForMaskedLM",),   TODO: need to tie weights
+            # ("AlbertForPretraining",),   TODO: need to tie weights
+            ("AlbertForMultipleChoice",),
+            # ("AlbertForQuestionAnswering",), TODO: transformers NOT add the last pool layer before qa_outputs
```
Comment (wj-Mcat) [inline tests/transformers/albert/test_modeling.py]: 可通过 architectures 来控制 pooler 的映射。

File: tests/transformers/albert/test_modeling.py [hunk 3]
```diff
@@ -337,6 +340,148 @@ def test_model_from_pretrained(self):
...
+                    torch_logit.detach().cpu().reshape([-1])[:9].numpy(),
+                    rtol=1e-4,
+                )
+            )
+
+    @parameterized.expand(
+        [
+            ("AlbertModel",),
+            # ("AlbertForMaskedLM",),   TODO: need to tie weights
+            # ("AlbertForPretraining",),   TODO: need to tie weights
```
Comment (wj-Mcat) [inline tests/transformers/albert/test_modeling.py:430]: 等 https://github.com/PaddlePaddle/PaddleNLP/pull/5623  合入之后就可以用 tie_weights 了。

**Review by wj-Mcat — DISMISSED** [review]

**Comment by megemini** [comment]
1. 之前 ``AlbertForQuestionAnswering`` 不过测的原因找到了，不是 ``pool`` 层的问题，是之前 ``name mapping`` 写错了，已经修改好，并且单测通过。
2. 关于 ``add_pooling_layer``，这个参数在 ``transformers`` 代码里面是写死的：
  ``` python
class AlbertForQuestionAnswering(AlbertPreTrainedModel):
    _keys_to_ignore_on_load_unexpected = [r"pooler"]

    def __init__(self, config: AlbertConfig):
        super().__init__(config)
        self.num_labels = config.num_labels

        self.albert = AlbertModel(config, add_pooling_layer=False) # 注意这里
        self.qa_outputs = nn.Linear(config.hidden_size, config.num_labels)

        # Initialize weights and apply final processing
        self.post_init()  
  ```
  在 ``transformers`` 的 ``config.json`` 中也没有这个配置项：
  ``` json
{
  "_name_or_path": "tiny_models/albert/AlbertForQuestionAnswering",
  "architectures": [
    "AlbertForQuestionAnswering"
  ],
  "attention_probs_dropout_prob": 0.1,
  "bos_token_id": 2,
  "classifier_dropout_prob": 0.1,
  "embedding_size": 128,
  "eos_token_id": 3,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 36,
  "initializer_range": 0.02,
  "inner_group_num": 1,
  "intermediate_size": 37,
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 512,
  "model_type": "albert",
  "num_attention_heads": 6,
  "num_hidden_groups": 6,
  "num_hidden_layers": 6,
  "pad_token_id": 0,
  "position_embedding_type": "absolute",
  "torch_dtype": "float32",
  "transformers_version": "4.28.0.dev0",
  "type_vocab_size": 16,
  "vocab_size": 30000
}
  ```

  而在 ``paddlenlp.transformers`` 中是个配置项：
  ``` python
class AlbertModel(AlbertPretrainedModel):
    def __init__(self, config: AlbertConfig):
        ...
        if config.add_pooling_layer:  # 注意这里
            self.pooler = nn.Linear(config.hidden_size, config.hidden_size)
            self.pooler_activation = nn.Tanh()
        else:
            self.pooler = None
            self.pooler_activation = None

        self.init_weights()
  ```
  当用 ``from_pretrained`` 引入模型的时候，默认将 ``add_pooling_layer`` 设为 ``True``。

  这样一般用没啥问题，因为模型的 ``logits`` 输出不走 ``pool`` 层，问题是，这时候 ``pool`` 层其实是随机初始化的，如果有人要用这个输出，那么结果是不可预测的。

  我这里把 ``pool`` 层做了个判断：
  ``` python
        if config.add_pooling_layer:
            model_mappings.extend(
                [
                    ["pooler.weight", "pooler.weight", "transpose"],
                    ["pooler.bias", "pooler.bias"],
                ]
            )
  ```
  请帮忙看看这样处理行不行？

  谢谢！

**Comment by sijunhe** [comment]
`AlbertForMaskedLM` 与 `AlbertForPretraining` 可以暂时不需要，处理好`AlbertForQuestionAnswering`即可

**Comment by megemini** [comment]
> `AlbertForMaskedLM` 与 `AlbertForPretraining` 可以暂时不需要，处理好`AlbertForQuestionAnswering`即可

嗯 `AlbertForQuestionAnswering` 已经处理好了～

**Comment by sijunhe** [comment]
需要merge develop一下哈

**Comment by megemini** [comment]
> 需要merge develop一下哈

搞定，请评审，谢谢！

**Comment by sijunhe** [comment]
> 请帮忙看看这样处理行不行？
> 
> 谢谢！

没问题的

**Review by sijunhe — APPROVED** [review]
lgtm
